File: agenkit-java/src/main/java/io/agenkit/budget/BudgetLimiter.java
package io.agenkit.budget;

import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Agent wrapper that enforces a USD cost budget.
 */
public final class BudgetLimiter implements Agent {

    private final Agent inner;
    private final double maxBudgetUsd;
    private final CostTracker costTracker;

    public BudgetLimiter(Agent inner, double maxBudgetUsd, String model) {
        this.inner = inner;
        this.maxBudgetUsd = maxBudgetUsd;
        this.costTracker = new CostTracker(model);
    }

    @Override
    public String getName() { return inner.getName() + "[budget]"; }

    @Override
    public List<String> getCapabilities() { return inner.getCapabilities(); }

    @Override
    public CompletableFuture<Message> process(Message message) {
        if (costTracker.getTotalCostUsd() >= maxBudgetUsd) {
            return CompletableFuture.failedFuture(
                    new RuntimeException("budget exceeded: $" + String.format("%.4f", maxBudgetUsd)
                            + " limit, spent $" + String.format("%.4f", costTracker.getTotalCostUsd())));
        }

        int estimatedInputTokens = message.contentString().length() / 4;
        return inner.process(message).thenApply(response -> {
            int estimatedOutputTokens = response.contentString().length() / 4;
            costTracker.record(estimatedInputTokens, estimatedOutputTokens);
            return response
                    .withMetadata("cost_usd", costTracker.getTotalCostUsd())
                    .withMetadata("budget_remaining",
                            maxBudgetUsd - costTracker.getTotalCostUsd());
        });
    }

    public double getRemainingBudget() {
        return maxBudgetUsd - costTracker.getTotalCostUsd();
    }

    @Override
    public IntrospectionResult introspect() {
        Map<String, Object> state = new HashMap<>();
        state.put("maxBudgetUsd", maxBudgetUsd);
        state.put("spentUsd", costTracker.getTotalCostUsd());
        state.put("remainingUsd", getRemainingBudget());
        return new IntrospectionResult(getName(), getCapabilities(), null, state, null);
    }
}

File: agenkit-cs/src/Agenkit/Budget/BudgetLimiter.cs
using Agenkit.Core;

namespace Agenkit.Budget;

/// <summary>
/// Wraps an agent and throws when the cost budget is exceeded.
/// </summary>
public class BudgetLimiter : IAgent
{
    private readonly IAgent _inner;
    private readonly CostTracker _tracker;
    private readonly decimal _budgetUsd;
    private readonly Func<Message, (long input, long output)> _tokenEstimator;

    /// <inheritdoc />
    public string Name => _inner.Name;

    /// <inheritdoc />
    public IReadOnlyList<string> Capabilities => _inner.Capabilities;

    /// <summary>Remaining budget in USD.</summary>
    public decimal RemainingBudget => _budgetUsd - _tracker.TotalCostUsd;

    /// <summary>Creates a BudgetLimiter.</summary>
    /// <param name="inner">Agent to wrap.</param>
    /// <param name="tracker">Cost tracker instance.</param>
    /// <param name="budgetUsd">Maximum allowed spend in USD.</param>
    /// <param name="tokenEstimator">Optional function to estimate token counts. Defaults to simple word count.</param>
    public BudgetLimiter(
        IAgent inner,
        CostTracker tracker,
        decimal budgetUsd,
        Func<Message, (long input, long output)>? tokenEstimator = null)
    {
        if (budgetUsd <= 0)
            throw new ArgumentOutOfRangeException(nameof(budgetUsd), "budget must be positive");
        _inner = inner;
        _tracker = tracker;
        _budgetUsd = budgetUsd;
        _tokenEstimator = tokenEstimator ?? DefaultEstimator;
    }

    /// <inheritdoc />
    public async Task<Message> ProcessAsync(Message message, CancellationToken ct = default)
    {
        if (_tracker.TotalCostUsd >= _budgetUsd)
            throw new InvalidOperationException(
                $"budget of ${_budgetUsd:F4} USD has been exhausted (spent: ${_tracker.TotalCostUsd:F4})");

        var response = await _inner.ProcessAsync(message, ct).ConfigureAwait(false);
        var (input, output) = _tokenEstimator(message);
        _tracker.Record(input, output);
        return response;
    }

    /// <inheritdoc />
    public IntrospectionResult Introspect() => new(
        Name,
        _inner.Capabilities,
        State: new Dictionary<string, object>
        {
            ["budget_usd"] = _budgetUsd,
            ["spent_usd"] = _tracker.TotalCostUsd,
            ["remaining_usd"] = RemainingBudget
        });

    private static (long input, long output) DefaultEstimator(Message m)
    {
        // Rough estimate: 1 token ≈ 4 characters
        var inputTokens = Math.Max(1, m.ContentString().Length / 4);
        return (inputTokens, inputTokens / 2);
    }
}
